### DISCUSSION TITLE
With thread-pool of 1 thread, can I use `_st`-sinks?

### DISCUSSION THREAD

**Discussion by raphaelzulliger** [gabime/spdlog#1852] [Q&A] [answered]
I have a multi-threaded application (Linux), some of it's threads have real-time priority (RT Kernel patch in place). I intend to

- Create several loggers (one logger per "topic", where a topic is something like "communication with X", "communication with Y", "general", ...)
- I intend to use async logging for all loggers (the intention is to minimize jitter on the "application path" as much as possible)
- I intend to use `stdout` and `stderr` sinks (finally, using `systemd`/`journald` to pass logs on)

I have not checked the implementation details of SpdLog, but from a logical standpoint, I would have assumed that with a thread pool of 1 thread, the sink(s) will be accessed by that single thread only and therefore I could use the "single threaded sinks" (`_st` versions). Right?

Is my thinking correct? Or more broadly: Is my approach of using single-threaded-thread-pool to decouple the "application path" from the "logging path" ok?

**Comment by gabime** [comment, marked as answer]
> .. pool of 1 thread, the sink(s) will be accessed by that single thread only and therefore I could use the "single threaded sinks" (_st versions). Right?

Right. As long as all loggers with the shared sinks use async logging.

> Is my thinking correct? Or more broadly: Is my approach of using single-threaded-thread-pool to decouple the "application path" from the "logging path" o

Sounds right. Please see [full queue policies](https://github.com/gabime/spdlog/wiki/6.-Asynchronous-logging#full-queue-policy) to choose how to handle cases when the queue gets full.

**Reply by raphaelzulliger** [reply]
Thanks for the quick/good response, I appreciate!

**Reply by raphaelzulliger** [reply]
IMPORTANT: Against what has been discussed above, the `_st` version seem *not to work* properly in multi-threaded environments!

Today, we had a situation in which much logs were generated and we ended up in application crashes.
We immediately changed from the `_st` versions to the `_mt` versions and the issue disappeared.
The "crash" happened in the `else` case of the following code:
```
// send the log message to the thread pool
SPDLOG_INLINE void spdlog::async_logger::sink_it_(const details::log_msg &msg)
{
    if (auto pool_ptr = thread_pool_.lock())
    {
        pool_ptr->post_log(shared_from_this(), msg, overflow_policy_);
    }
    else
    {
        throw_spdlog_ex("async log: thread pool doesn't exist anymore");
    }
}
```
We did not spend much time understanding the issue, but the thread pool still existed (thus, the error message is kind of misleading). Admittedly, it's not fully proofed that the change to `_mt` really fixed the issue, but so far we had no more crashes while before, we could reproduce the issue by 100%.

**Comment by gabime** [comment]
Strange.. I don't understand how “_st/_mt” sinks can affect to the lifetime of the thread pool.  How do you create the thread pool and loggers? can you share ?

**Reply by software-natives-docker** [reply]
Sure, I can share (a very minimally modified version):
- SetupSinks is called once during early startup
- CreateLogger is called about 10 times to create 10 different loggers, used by 10 different threads

```

void SetupSinks()
{
  spdlog::init_thread_pool(8192, 1);
  m_pStandardOutputSink = std::make_shared<spdlog::sinks::stdout_color_sink_st >();

  auto nMaxFileSize = 10 * 1024 * 1024; // 10 MBytes
  auto nMaxFileCount = 3;
  m_pFileOutputSink = std::make_shared<spdlog::sinks::rotating_file_sink_st >("XXXXX", nMaxFileSize, nMaxFileCount);

  spdlog::flush_every(std::chrono::seconds(5));
}


std::shared_ptr<spdlog::logger> CreateLogger(const std::string& i_LoggerName)
{
  std::vector<spdlog::sink_ptr> sinks{m_pStandardOutputSink};
#ifndef NDEBUG
  // in "Debug builds", we output to a log file always.
  sinks.push_back(m_pFileOutputSink);
#endif
  auto pLogger = std::make_shared<spdlog::async_logger>(
      i_LoggerName, sinks.begin(), sinks.end(), spdlog::thread_pool(), spdlog::async_overflow_policy::overrun_oldest);

  const auto level = ...
  pLogger->set_level(level);

  pLogger->set_pattern("[%Y-%m-%d %T.%e] [%n] [%^%l%$] [%s:%# %!] %v");

  return pLogger;
}
```

**Comment by gabime** [comment]
> SetupSinks is called once during early startup

Does it happen in some static init section before main? Or perhaps the CreateLogger() calls? this might cause using different thread pools due to the undefined init order of statics.

Also, spdlog’s stdout sinks all share same global console mutex, so there a chance  that there is a use of stdout logger elsewhere(e.g. call to spdlog::info(..))?

btw the `spdlog::flush_every(std::chrono::seconds(5));` won’t work if the loggers are not registered(call `register_logger(..)`)

**Reply by software-natives-docker** [reply]
Interesting. The saying "You can't solve a problem you did not yet understand" seems to have hit me her

I do first use `spdlog` at the very beginning of `main()` (definitely no static/global variable's CTOR or the like.

Since we moved from a "legacy logger" to `spdlog`, I can say that nobody is calling `spdlog::info()` (or any of the other functions) directly. I also searched the whole project for such function calls

BUT: I just realized that my system ends up with two `spdlog`-related threads:
- the first is created upon a call to `spdlog::init_thread_pool`
- the second is created on the call to `spdlog::flush_every`

Thanks for the hint of the useless call to the `spdlog::flush_every`! But does it make sense to cause issues like I'm having here?

**Reply by software-natives-docker** [reply]
I've created the following "stress test" that should resemble our situation:

```c++
#include <spdlog/spdlog.h>
#include <spdlog/async.h>
#include <spdlog/sinks/stdout_color_sinks.h>
#include <spdlog/sinks/rotating_file_sink.h>

#include <gtest/gtest.h>
#include <thread>
#include <future>

class Logger {
public:
    Logger() {
        SetupSinks();
    }
    void SetupSinks()
    {
        spdlog::init_thread_pool(8192, 1);
        m_pStandardOutputSink = std::make_shared<spdlog::sinks::stdout_color_sink_st >();

        auto nMaxFileSize = 10 * 1024 * 1024; // 10 MBytes
        auto nMaxFileCount = 3;
        m_pFileOutputSink = std::make_shared<spdlog::sinks::rotating_file_sink_st >("/tmp/XXX", nMaxFileSize, nMaxFileCount);

        spdlog::flush_every(std::chrono::seconds(5));
    }


    std::shared_ptr<spdlog::logger> CreateLogger(const std::string& i_LoggerName)
    {
        std::vector<spdlog::sink_ptr> sinks{m_pStandardOutputSink};
        #ifndef NDEBUG
        // in "Debug builds", we output to a log file always.
        sinks.push_back(m_pFileOutputSink);
        #endif
        auto pLogger = std::make_shared<spdlog::async_logger>(
            i_LoggerName, sinks.begin(), sinks.end(), spdlog::thread_pool(), spdlog::async_overflow_policy::overrun_oldest);

        // const auto level = ...
        // pLogger->set_level(level);

        pLogger->set_pattern("[%Y-%m-%d %T.%e] [%n] [%^%l%$] [%s:%# %!] %v");

        return pLogger;
    }
private:

    std::shared_ptr<spdlog::sinks::sink> m_pStandardOutputSink;
    std::shared_ptr<spdlog::sinks::sink> m_pFileOutputSink;
};


class SpdLogStressTest : public testing::Test {
};

TEST_F(SpdLogStressTest, DefaultValues_NotRunning)
{
    Logger log;

    const size_t cLogCnt = 100000000;

    auto t1 = std::thread([&]() {
            auto logger = log.CreateLogger("Logger1");
            for (size_t i=0; i<cLogCnt; ++i) {
                logger->warn("Logger1 {}", i);
            }
        });
    auto t2 = std::thread([&]() {
            auto logger = log.CreateLogger("Logger2");
            for (size_t i=0; i<cLogCnt; ++i) {
                logger->warn("Logger2 {}", i);
            }
        });

    t1.join();
    t2.join();
}
```
Unfortunately, the issue does not happen as in our real application. But "sometimes" I run into the following crash with that code (but I assume it's because of the call to `set_pattern` and thus yet another potential issue in my code)
```
spdlog::pattern_formatter::format(spdlog::pattern_formatter * const this, const spdlog::details::log_msg & msg, spdlog::memory_buf_t & dest) (/home/raphael/.conan/data/spdlog/1.8.2/_/_/build/93af92fd044bc44291351609be0c2aa41667b1e7/source_subfolder/include/spdlog/pattern_formatter-inl.h:1039)
spdlog::sinks::ansicolor_sink<spdlog::details::console_nullmutex>::log(spdlog::sinks::ansicolor_sink<spdlog::details::console_nullmutex> * const this, const spdlog::details::log_msg & msg) (/home/raphael/.conan/data/spdlog/1.8.2/_/_/build/93af92fd044bc44291351609be0c2aa41667b1e7/source_subfolder/include/spdlog/sinks/ansicolor_sink-inl.h:49)
spdlog::async_logger::backend_sink_it_(spdlog::async_logger * const this, const spdlog::details::log_msg & msg) (/home/raphael/.conan/data/spdlog/1.8.2/_/_/build/93af92fd044bc44291351609be0c2aa41667b1e7/source_subfolder/include/spdlog/async_logger-inl.h:63)
spdlog::details::thread_pool::process_next_msg_(spdlog::details::thread_pool * const this) (/home/raphael/.conan/data/spdlog/1.8.2/_/_/build/93af92fd044bc44291351609be0c2aa41667b1e7/source_subfolder/include/spdlog/details/thread_pool-inl.h:108)
spdlog::details::thread_pool::worker_loop_(spdlog::details::thread_pool * const this) (/home/raphael/.conan/data/spdlog/1.8.2/_/_/build/93af92fd044bc44291351609be0c2aa41667b1e7/source_subfolder/include/spdlog/details/thread_pool-inl.h:90)
spdlog::details::thread_pool::<lambda()>::operator()(void) const(const spdlog::details::thread_pool::<lambda()> * const __closure) (/home/raphael/.conan/data/spdlog/1.8.2/_/_/build/93af92fd044bc44291351609be0c2aa41667b1e7/source_subfolder/include/spdlog/details/thread_pool-inl.h:28)
std::__invoke_impl<void, spdlog::details::thread_pool::thread_pool(size_t, size_t, std::function<void()>)::<lambda()> >(std::__invoke_other, spdlog::details::thread_pool::<lambda()> &&)(spdlog::details::thread_pool::<lambda()> && __f) (/usr/include/c++/9/bits/invoke.h:60)
std::__invoke<spdlog::details::thread_pool::thread_pool(size_t, size_t, std::function<void()>)::<lambda()> >(spdlog::details::thread_pool::<lambda()> &&)(spdlog::details::thread_pool::<lambda()> && __fn) (/usr/include/c++/9/bits/invoke.h:95)
std::thread::_Invoker<std::tuple<spdlog::details::thread_pool::thread_pool(size_t, size_t, std::function<void()>)::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>)(std::thread::_Invoker<std::tuple<spdlog::details::thread_pool::thread_pool(size_t, size_t, std::function<void()>)::<lambda()> > > * const this) (/usr/include/c++/9/thread:244)
std::thread::_Invoker<std::tuple<spdlog::details::thread_pool::thread_pool(size_t, size_t, std::function<void()>)::<lambda()> > >::operator()(void)(std::thread::_Invoker<std::tuple<spdlog::details::thread_pool::thread_pool(size_t, size_t, std::function<void()>)::<lambda()> > > * const this) (/usr/include/c++/9/thread:251)
std::thread::_State_impl<std::thread::_Invoker<std::tuple<spdlog::details::thread_pool::thread_pool(size_t, size_t, std::function<void()>)::<lambda()> > > >::_M_run(void)(std::thread::_State_impl<std::thread::_Invoker<std::tuple<spdlog::details::thread_pool::thread_pool(size_t, size_t, std::function<void()>)::<lambda()> > > > * const this) (/usr/include/c++/9/thread:195)
libstdc++.so.6![Unknown/Just-In-Time compiled code] (Unknown Source:0)
libpthread.so.0!start_thread(void * arg) (/build/glibc-eX1tMB/glibc-2.31/nptl/pthread_create.c:477)
libc.so.6!clone() (/build/glibc-eX1tMB/glibc-2.31/sysdeps/unix/sysv/linux/x86_64/clone.S:95)
```

On the following line of code (see `--->`) since `f` is a `nulptr`
```c++

SPDLOG_INLINE void pattern_formatter::format(const details::log_msg &msg, memory_buf_t &dest)
{
    auto secs = std::chrono::duration_cast<std::chrono::seconds>(msg.time.time_since_epoch());
    if (secs != last_log_secs_)
    {
        cached_tm_ = get_time_(msg);
        last_log_secs_ = secs;
    }

    for (auto &f : formatters_)
    {
--->        f->format(msg, cached_tm_, dest);
    }
    // write eol
    details::fmt_helper::append_string_view(eol_, dest);
}
```

**Reply by software-natives-docker** [reply]
Does that all make sense to you? Should it also resolve if I remove the call to `spdlog::flush_every(std::chrono::seconds(5));`

**Reply by software-natives-docker** [reply]
I could not reproduce this `pattern`-crash anymore when setting the pattern on the sink (which makes much more sense anyway). Thanks for this hint.

Any hypothesis on why this test (as well as our application) ends up with two "spdlog" threads?

**Comment by gabime** [comment]
> Unfortunately, the issue does not happen as in our real application. But "sometimes" I run into the following crash with that code (but I assume it's because of the call to set_pattern and thus yet another potential issue in my code)

Right,  set_pattern [propagates](https://github.com/gabime/spdlog/blob/be14e60d9e8be31735dd9d2d132d8a4cd3482165/include/spdlog/logger-inl.h#L91 ) to the sink, so no wonder it crashes if called from 2 threads(and the sink is _st).  You can call  `m_pFileOutputSink->set_pattern(..)` once in the SetupSinks() func.

Would be good if you fix and try again to make sure the test indeed passes as expected.

> Should it also resolve if I remove the call to spdlog::flush_every(std::chrono::seconds(5));

It has no effect since the loggers are not registered.

**Comment by gabime** [comment]
> Any hypothesis on why this test (as well as our application) ends up with two "spdlog" threads

Yes. seems normal. one for the thread pool and one for the flush_every() thread. Should not cause crash.

**Reply by software-natives-docker** [reply]
I'll improve/fix our applications and ask my colleague, whether he may try to reproduce the initially reported crash using the `_st` version (he had a setup in which could perfectly reproduce, I don't have that. It's an embedded system setup).

Many thanks for the great help!

**Comment by gabime** [comment]
ok..feel free to update on any news.
